newrelic/hooks/mlmodel_openai.py

```diff
@@ -0,0 +1,145 @@
+
+# Copyright 2010 New Relic, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import openai
+import uuid
+from newrelic.api.function_trace import FunctionTrace
+from newrelic.common.object_wrapper import wrap_function_wrapper
+from newrelic.api.transaction import current_transaction
+from newrelic.api.time_trace import get_trace_linking_metadata
+from newrelic.core.config import global_settings
+from newrelic.common.object_names import callable_name
+from newrelic.core.attribute import MAX_LOG_MESSAGE_LENGTH
+
+
+def wrap_chat_completion_create(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+
+    if not transaction:
+        return
+
+    ft_name = callable_name(wrapped)
+    with FunctionTrace(ft_name) as ft:
+        response = wrapped(*args, **kwargs)
+
+    if not response:
+        return
+
+    custom_attrs_dict = transaction._custom_params
+    conversation_id = custom_attrs_dict.get("conversation_id", str(uuid.uuid4()))
+
+    chat_completion_id = str(uuid.uuid4())
+    available_metadata = get_trace_linking_metadata()
+    span_id = available_metadata.get("span.id", "")
+    trace_id = available_metadata.get("trace.id", "")
+
+    response_headers = getattr(response, "_nr_response_headers", None)
+    response_model = response.model
+    response_id = response.get("id", "")
+    request_id = response_headers.get("x-request-id", "")
+    settings = transaction.settings if transaction.settings is not None else global_settings()
+
+    chat_completion_summary_dict = {
+        "id": chat_completion_id,
+        "appName": settings.app_name,
+        "conversation_id": conversation_id,
+        "span_id": span_id,
+        "trace_id": trace_id,
+        "transaction_id": transaction._transaction_id,
+        "request_id": request_id,
```

Comment:
This wasn't in the open-ai CDD though we need it for feedback. Scope wise, it makes sense to attach it here. But it's only used for identifying a message in the feedback CDD so I wonder if we need it here or just on the LlmMessage event? I suppose it doesn't hurt to have both, it's a pretty small amount of data.

Alternatively, we could make the chat completion id the request id just like we made the message id the response id + the index? Maybe we should ask the group about this?

Reply:
Good point! I checked about this and I think we're going to go ahead and keep the request ID on the summary event and leave the chat completion ID as a UUID for our first iteration.